backend/src/db.js

```diff
@@ -18,93 +19,92 @@ export class User {
-        }
     }
 
-    upvotePost(data) {
-        const voter = this.users.find(user => user.name === data.voter.name)
-        if (voter) {
-            const post = this.posts.find(post => post.id === data.id)
-            if (post) {
-                const userVotedPosts = post.votes.find(u => u.name === data.voter.name)
-                if (!userVotedPosts) {
-                    const newPosts = this.posts.map(p => p.id === data.id ? this.addVoterToPost(p, voter) : p)
-                    this.posts = [...newPosts]
-                    return this.posts.find(p => p.id === data.id)
-                } else throw new UserInputError("User already voted")
-            } else throw new UserInputError("Post does not exist")
-        } else throw new AuthenticationError("User does not exist")
+    signupUser(data) {
+        if (data.password.length < 8) return new Error("Password must have at least 8 characters")
+
+        const user = this.users.find(u => u.email === data.email)
+        if (user !== undefined) return new Error("Email already exist")
+
+
+        const encryptedPassword = bcrypt.hashSync(data.password, this.salt)
+        const newUser = new User({
+            name: data.name,
+            email: data.email,
+            password: encryptedPassword,
+        })
+        this.users.push(newUser)
+
+        return jwt.sign(
+            {id: newUser.id},
+            this.secret)
     }
 
-    getVotes(postId) {
-        return this.posts.find(post => post.id === postId).votes.length
+    verifyToken(token) {
+        try {
+            return jwt.verify(token, this.secret)
+        } catch (err) {
+            return undefined
+        }
     }
 
-    addVoterToPost(post, voter) {
-        post.votes.push(voter)
-        return post
+    createPost(data, context) {
+        const author = this.users.find(user => user.id === context.user.id)
+        if (author === undefined) return new Error("User does not exist")
+
+        const newPost = new Post({
+            title: data.post.title,
+            author: author
+        })
+        this.posts.push(newPost)
+
+        return newPost
     }
 
-    addPostToAuthor(post, author) {
-        author.posts.push(post)
-        return author
+    upvotePost(data, context) {
+        const voter = this.users.find(user => user.id === context.user.id)
+        if (voter === undefined) return new Error("User does not exist")
+
+        const post = this.posts.find(post => post.id === data.id)
+        if (post === undefined) return new Error("Post does not exist")
+
+        const userVotedPosts = post.votes.find(u => u.id === voter.id)
+        if (userVotedPosts !== undefined) {
+            return new Error("User already voted")
+        } else {
+            const newPosts = this.posts.map(p => p.id === data.id ? this.addVoterToPost(p, voter) : p)
+            this.posts = [...newPosts]
+            return this.posts.find(p => p.id === data.id)
+        }
     }
```

Comment:
:star: For assigning the   authenticated user to a post in write and upvote mutations.